src/operations/aggregate.ts

```diff
@@ -79,6 +79,10 @@ export class AggregateOperation<T = Document> extends CommandOperation<T> {
     }
   }
 
+  get isCursorCreating(): boolean {
```

Comment:
This seems it would be better suited to be an [aspect](https://github.com/mongodb/node-mongodb-native/blob/4.0/src/operations/aggregate.ts#L130). If we could refactor to that it would be great but also something we can defer as well.

Reply:
Do we actually need this on operations classes? The cursor design is such that subclasses of `AbstractCursor` implement their own initialization which execute an operation, so we _should_ have all the context we need about which operations create cursors without explicitly adding that context to operations classes. Is there a reason you can't move the check in `executeOperation` into the `AbstractCursor` call to [_initialize](https://github.com/mongodb/node-mongodb-native/blob/4.0/src/cursor/abstract_cursor.ts#L686) ?

Reply:
Well the addition of this was specifically for retry behaviour, and I wanted to keep all the retry related code in the same place and not have it scattered in multiple places.

Reply:
Also because of the `findOne` case - it doesn't directly instantiate a cursor, it calls `collection.find` in it's operation which means that the session in that case it not owned by the cursor as `executeOperation` creates the implicit session, not `cursor._initialize`.

Reply:
Refactored in a026ce4ca2d7dd337f4b13b0cac09d7a813a6924 to use aspects.

Reply:
Hm, but I think this actually complicates the retry behavior. The logic in `executeOperation` now is very clear: if the operation is retryable, and the error is retryable, the operation is retried. Now we have a special logical block that has specific knowledge about cursors, and I'm worried that we will continue to lump extraneous checks there. 

I'm not sure I follow the `findOne` case, can you explain that a bit more? It should be fine that it calls another operation internally. In fact, I think marking `FindOneOperation` as being  "cursor creating" is incorrect: it issues a command which returns a single-batch, exhausted cursor. The session should not have a pinned connection after that completes, and if it does that would mean we have a bug somewhere else. Unfortunately, we'd never see that now because we put this aspect on the operation, and that means in `executeOperation` we are just force unpinning, and force resetting the pool. Seems dangerous.

Reply:
What tests were failing without this condition? I tried running the tests locally without the creates_cursor aspect and all passed. There are some skips so maybe my environment is incomplete. (I have both multi and single env variables)

Reply:
No tests fail without this condition, but it leaks the connection each time `findOne` retries. This is because `findOne` calls `executeOperation`, which itself creates an implicit session with the `owner` being a symbol. Then inside it's execute method it calls `collection.find`, which creates a cursor with the already created implicit session. The connection is pinned at the connection pool level because `find` is a pinnable operation, and then when the cursor exhausts it does not unpin because the cursor is not the owner of the session. So the connection remains pinned to the session.

Reply:
<img width="1904" alt="Screenshot 2021-08-03 at 16 18 31" src="https://user-images.githubusercontent.com/9030/128031554-4f932f2f-7df6-4c42-812b-89cb7dc70c3a.png">

Reply:
You can see at the end of the test run here without this code that there are 2 sessions still open to the load balancer, which correlated directly with 2 findOne retries that had an extra connection checked out of the pool when the client was closed.

Reply:
<img width="1904" alt="Screenshot 2021-08-03 at 16 18 31" src="https://user-images.githubusercontent.com/9030/128032143-779f355f-c87b-4ef4-ab1c-d940f8c2c66b.png">

Here are the extra 2 checked out connections from the findOne retries at the end of the test run.

Reply:
I do agree in this case that the pin should never have occurred, but until we refactor the operation layer it's always going to be pinned here because we pin on connection checkout from the pool and the find command is pinnable.